function ManhattanPlink( filename, sex )

%This function takes a file outputted by plink and plots a Manhattan Plot.
%Tested using assoc, assoc.logistic and assoc.fisher files generated by
%Plink 1.7. By default, sex chromoromes are not outputted. The output is
%stored in a publication-ready png format and an editable .fig format.
%
%   Usage: ManhattanPlink('gwas.assoc.fisher') will take the association
%   analysis in 'gwas.assoc.fisher, generate a Manhattan Plot, and store it
%   in gwas_ManhattanPlot.png, which should be publication-ready, and
%   gwas_ManhattanPlot.fig for minor readjustments in MATLAB's GUI.

%   ManhattanPlink('gwas.assoc.fisher',1) specifies that the sex
%   chromosomes should be plotted.
%
%   Harry Green (2018) Genetics of Complex Traits

figure()
if exist('sex')==0
    sex=0;
end

%MATLAB refuses to read file formats like .assoc, so first rename as .txt
copyfile(filename, strcat(filename,'.txt'));
opts = detectImportOptions( strcat(filename,'.txt'),'NumHeaderLines',0);
T = readtable(strcat(filename,'.txt'),opts); % the columns of T will be the headers of the assoc file
delete(strcat(filename,'.txt')) %delete unwanted .txt file that we didn't want anyway

tab2=[T.CHR,T.BP,T.P]; % These are the only useful columns for a manhattan plot, chromosome, base pair, p value
if sex==0
    tab2=tab2(tab2(:,1)<23,:); %remove chr23 if not wanting sex chromosomes
end

bptrack=0; %variable to track base pairs, this helps the plotting function to know where to start plotting the next chromosome
tab2(tab2(:,3)==0,:)=[];
for i=1:22+sex
    hold on
    plot(tab2(tab2(:,1)==i,2)+max(bptrack),-log10(tab2(tab2(:,1)==i,3)),'.'); %a scatterplot. On the x axis, the base pair number + bptrack, which starts at 0. On the y axis, - log 10 p
    bptrack=[bptrack,max(bptrack)+max(max(tab2(tab2(:,1)==i,:)))]; %this updates bptrack by adding the highest base pair number of the chromosome. At the end, we should get the total number of base pairs in the human genome. All values of bptrack are stored
end

ylimit=max(max(-log10(tab2(:,3)))+1); %if strongest hit is 1e-60, plot window goes up to 1e-61
ylimitmin=min(min(-log10(tab2(:,3))));

plot([0,max(bptrack)],-log10([5e-8,5e-8]),'k') %genome wide significant line

xlabel('Chromosome','Interpreter','latex')
ylabel('$-\log_{10}(p)$','Interpreter','latex')
xlim([0,max(bptrack)])
ylim([ylimitmin,max(ylimit,-log10(1e-8))]) %y axis will always go to 1e-8 at least

M=movmean(bptrack,2); %this calculates the moving average of bptrack and uses them as chromosome label markers.
xticks(M(2:end));
xticklabels( 1:23 );

a = strsplit(filename,'.');

box on
title('Manhattan Plot','Interpreter','latex')
name=[a{1},'_ManhattanPlot.fig'];
savefig(name);

set(gcf, 'paperunits', 'centimeters', 'paperposition', [0 0 30 15])
print([a{1},'_ManhattanPlot.png'],'-dpng')
end

